server/etcdserver/raft.go

```diff
@@ -314,6 +328,38 @@ func (r *raftNode) start(rh *raftReadyHandler) {
 	}()
 }
 
+// For a cluster with only one member, the raft always send identical
+// unstable entries and committed entries to etcdserver, and etcd
+// responds to the client once it finishes (actually partially) the
+// applying workflow.
+//
+// When the client receives the response, it doesn't mean etcd has already
+// successfully saved the data, including BoltDB and WAL, because:
+//    1. etcd commits the boltDB transaction periodically instead of on each request;
+//    2. etcd saves WAL entries in parallel with applying the committed entries.
+// Accordingly, it might run into a situation of data loss when the etcd crashes
+// immediately after responding to the client and before the boltDB and WAL
+// successfully save the data to disk.
+// Note that this issue can only happen for clusters with only one member.
+//
+// For clusters with multiple members, it isn't an issue, because etcd will
+// not commit & apply the data before it being replicated to majority members.
+// When the client receives the response, it means the data must have been applied.
+// It further means the data must have been committed.
+// Note: for clusters with multiple members, the raft will never send identical
+// unstable entries and committed entries to etcdserver.
+func shouldWaitWALSync(unstableEntries []raftpb.Entry, committedEntries []raftpb.Entry) bool {
```

Comment:
Is there a way to have a single code path that is safe regardless of whether we're in multi-server or single server mode?

Reply:
I do not get your point. 

For multi-member cluster, there is no need to wait for the WAL sync, and this function will always return `false`.

Reply:
There are two solutions in my mind before delivering this PR.

The first solution is to enhance the existing raft protocal. The existing raft workflow commit each log immediately when it receves each proposal for clusters with only member, because it doesn't need to get confirmation from itself. Accordingly it sends both the unstable logs and committed logs to etcdserver. The solution is to send a message to etcdserver and wait for the confirmation, no matter it's single-server or multi-server. The good side of this solution is that it looks elegant. The bad side it has some impact on the performance, and it also needs to update the stable raft package. It might be what your `a single code path` means.

The second solutio is what this PR delivers. The good side is that it has little performance impact, and no impact on multi-server clusters at all. The bad side is that it complicates the applying workflow, but it should be accepted. 

Eventually I followed the second solution above for now.

Reply:
Proposed fix by @ahrtr makes sense for me.